hbase-client/src/main/java/org/apache/hadoop/hbase/client/ConnectionUtils.java

```diff
@@ -695,4 +606,94 @@ static void updateStats(Optional<ServerStatisticTracker> optStats,
         metrics -> ResultStatsUtil.updateStats(metrics, serverName, regionName, regionLoadStats));
     });
   }
+
+  @FunctionalInterface
+  interface Converter<D, I, S> {
+    D convert(I info, S src) throws IOException;
+  }
+
+  @FunctionalInterface
+  interface RpcCall<RESP, REQ> {
+    void call(ClientService.Interface stub, HBaseRpcController controller, REQ req,
+        RpcCallback<RESP> done);
+  }
+
+  static <REQ, PREQ, PRESP, RESP> CompletableFuture<RESP> call(HBaseRpcController controller,
+      HRegionLocation loc, ClientService.Interface stub, REQ req,
+      Converter<PREQ, byte[], REQ> reqConvert, RpcCall<PRESP, PREQ> rpcCall,
+      Converter<RESP, HBaseRpcController, PRESP> respConverter) {
+    CompletableFuture<RESP> future = new CompletableFuture<>();
+    try {
+      rpcCall.call(stub, controller, reqConvert.convert(loc.getRegion().getRegionName(), req),
+        new RpcCallback<PRESP>() {
+
+          @Override
+          public void run(PRESP resp) {
+            if (controller.failed()) {
+              future.completeExceptionally(controller.getFailed());
+            } else {
+              try {
+                future.complete(respConverter.convert(controller, resp));
+              } catch (IOException e) {
+                future.completeExceptionally(e);
+              }
+            }
+          }
+        });
+    } catch (IOException e) {
+      future.completeExceptionally(e);
+    }
+    return future;
+  }
+
+  static ThreadPoolExecutor getThreadPool(Configuration conf, int maxThreads, int coreThreads,
```

Comment:
This pool is shared by what in the client?

Reply:
Or it is just generic utility? Want to say in comment what choices it makes internally when say maxThreads == 0.

Reply:
Shared inside a Connection. It is used for implementation coprocessor calls for sync client, as the old method allow users to reference a blocking method in the protobuf stub, so we must run them in a separated thread...

Reply:
Does this need more comments saying that its for CPs only? Or maybe there are comments to this effect elsewhere, just not down here at this level?

Reply:
I moved this method to ConnectionOverAsyncConnection in HBASE-22550. But anyway, we could add comments here. Let me do some hack here to amend the comments in the commit for HBASE-22550 since we are on a feature branch.